draft-ietf-masque-connect-udp-listen.md

```diff
@@ -154,15 +176,118 @@ UDP Payload:
+the given Context ID (For example, due to considerable memory requirements of
+establishing  a list of mappings per target per client). If the compression was
+rejected, the client and proxy MUST use an uncompressed context ID (See {{uncompressed}}) to exhange
+UDP payloads for the given target.
+
+### Uncompressed datagrams {#uncompressed}
+
+If the client wishes to allocate a Context ID for uncompressed packets,
+it MUST first exchange the COMPRESSION_ASSIGN capsule (see {{capsuleassignformat}})
+with the proxy with an unused Context ID defined in {{contextid}} with
+the IP Length set to zero.
 
-The "Connect-UDP-Bind" header field’s value is an Integer. It is set as the
-Context ID allocated for Bound UDP Proxying; see {{mechanism}}. Any other
-value type MUST be handled as if the field were not present by the recipients
-(for example, if this field is defined multiple times, its type becomes a List
-and therefore is to be ignored). This document does not define any parameters
-for the Connect-UDP-Bind header field value, but future documents might define
-parameters. Receivers MUST ignore unknown parameters.
+
+### Compression Mapping {#mappings}
+
+When an endpoint receives a COMPRESSION_ASSIGN capsule with a non-zero IP length, it MUST decide whether to accept or reject the compression mapping:
+
+if it accepts the mapping, first the receiver MUST save the mapping from context ID to address and port. Second, the receiver MUST echo an identical COMPRESSION_ASSIGN capsule back to its peer.
+
+if it rejects the mapping, the receiver MUST respond by sending a COMPRESSION_CLOSE capsule with the context ID set to the one from the received COMPRESSION_ASSIGN capsule
+
+The client or proxy MAY choose to close any context contexts that it registered
+or was registered with it respectively using COMPRESSION_CLOSE
+(For example when a mapping is unused for a long time). Another potential use is
+{{restrictingips}}.
+
+
+## Restricting IPs {#restrictingips}
+
+If an uncompressed Context ID was set (via {{uncompressed}}), the client MAY at any point request the proxy reject all traffic from uncompressed
+targets by using COMPRESSION_CLOSE (see {{compressionclose}}) on said Context ID.
+targets effectively acting as a firewall against unwanted or unknown IPs.
+
+
+## Capsules {#capsules}
+The Listener capsule types are defined as follows:
+
+### The COMPRESSION_ASSIGN capsule {#compressionassign}
+
+The Compression Assign capsule has two purposes. Either to request the assignment of a Context ID (see {{contextid}}) to a corresponding target IP:Port. Or to accept a COMPRESSION_ASSIGN request from the other party.
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_ASSIGN (0x1F1F1F1F),
+  Length (i),
+  Target Information,
+}
+~~~
+{: #capsuleassignformat title="Compression Assign Capsule Format"}
+
+~~~ ascii-art
+Target Information {
+  Context ID (i),
+  IP Version (8),
+  IP Address (32..128),
+  UDP Port (16),
```

Comment:
We need to account for the IP version = 0 scenario by making the IP address and port optional

```suggestion
  [IP Address (32..128)],
  [UDP Port (16)],
```